p2p/discovery/mdns.go

```diff
@@ -37,186 +36,183 @@ type Notifee interface {
-	}
-	service, err := mdns.NewMDNSService(myid, serviceTag, "", "", port, ipaddrs, info)
+	addrs, err := peer.AddrInfoToP2pAddrs(&peer.AddrInfo{
+		ID:    s.host.ID(),
+		Addrs: interfaceAddrs,
+	})
 	if err != nil {
-		return nil, err
+		return err
 	}
-
-	// Create the mDNS server, defer shutdown
-	server, err := mdns.NewServer(&mdns.Config{Zone: service})
-	if err != nil {
-		return nil, err
+	var txts []string
+	for _, addr := range addrs {
+		if manet.IsThinWaist(addr) { // don't announce circuit addresses
+			txts = append(txts, dnsaddrPrefix+addr.String())
+		}
 	}
 
-	s := &mdnsService{
-		server:   server,
-		service:  service,
-		host:     peerhost,
-		interval: interval,
-		tag:      serviceTag,
+	ips, err := s.getIPs(addrs)
+	if err != nil {
+		return err
+	}
+
+	server, err := zeroconf.RegisterProxy(
+		s.mdnsInstance(),
+		s.serviceName,
+		mdnsDomain,
+		4001,
+		s.host.ID().Pretty(), // TODO: deals with peer IDs longer than 63 characters
+		ips,
+		txts,
+		nil,
+	)
+	if err != nil {
+		return err
 	}
-
-	go s.pollForEntries(ctx)
-
-	return s, nil
+	s.server = server
+	return nil
 }
 
-func (m *mdnsService) Close() error {
-	return m.server.Shutdown()
-}
-
-func (m *mdnsService) pollForEntries(ctx context.Context) {
-	ticker := time.NewTicker(m.interval)
-	defer ticker.Stop()
-
-	for {
-		//execute mdns query right away at method call and then with every tick
-		entriesCh := make(chan *mdns.ServiceEntry, 16)
-		go func() {
-			for entry := range entriesCh {
-				m.handleEntry(entry)
+func (s *mdnsService) startResolver() {
+	entryChan := make(chan *zeroconf.ServiceEntry, 1000)
+	go func() {
+		for entry := range entryChan {
+			// We only care about the TXT records.
+			// Ignore A, AAAA and PTR.
+			addrs := make([]ma.Multiaddr, 0, len(entry.Text)) // assume that all TXT records are dnsaddrs
+			for _, s := range entry.Text {
+				if !strings.HasPrefix(s, dnsaddrPrefix) {
+					log.Debug("missing dnsaddr prefix")
+					continue
+				}
+				addr, err := ma.NewMultiaddr(s[len(dnsaddrPrefix):])
+				if err != nil {
+					log.Debugf("failed to parse multiaddr: %s", err)
+					continue
+				}
+				addrs = append(addrs, addr)
 			}
-		}()
-
-		log.Debug("starting mdns query")
-		qp := &mdns.QueryParam{
-			Domain:  "local",
-			Entries: entriesCh,
-			Service: m.tag,
-			Timeout: time.Second * 5,
-		}
-
-		err := mdns.Query(qp)
-		if err != nil {
-			log.Warnw("mdns lookup error", "error", err)
+			infos, err := peer.AddrInfosFromP2pAddrs(addrs...)
+			if err != nil {
+				log.Debugf("failed to get peer info: %s", err)
+				continue
+			}
+			s.mutex.Lock()
+			for _, info := range infos {
+				for _, notif := range s.notifees {
+					go notif.HandlePeerFound(info)
+				}
+			}
+			s.mutex.Unlock()
 		}
-		close(entriesCh)
-		log.Debug("mdns query complete")
-
-		select {
-		case <-ticker.C:
-			continue
-		case <-ctx.Done():
-			log.Debug("mdns service halting")
-			return
+	}()
+	go func() {
+		defer close(s.resolverRunning)
```

Comment:
Consider using a waitgroup. We'd be able to reuse it in for both goroutines.

Reply:
Good point. Done.